plugin-server/src/worker/ingestion/event-property-counter.ts

```diff
@@ -0,0 +1,178 @@
+import { timeoutGuard } from '../../utils/db/utils'
+import { status } from '../../utils/status'
+
+enum EventPropertyType {
+    Number = 'NUMBER',
+    String = 'STRING',
+    Boolean = 'BOOLEAN',
+    DateTime = 'DATETIME',
+}
+
+interface EventPropertiesCounter {
+    createdAt: number
+    lastSeenAt: number
+    totalVolume: number
+    propertyType: EventPropertyType | null
+    propertyTypeFormat: string | null
+}
+
+interface EventNamePropertiesBuffer {
+    totalVolume: number
+    uniqueVolume: number
+    buffer: Map<TeamId, Map<string, EventPropertiesCounter>>
+}
+
+export class EventPropertyCounter {
+    db: DB
+    eventPropertiesBuffer: EventNamePropertiesBuffer
+    lastFlushAt: DateTime
+    statsd?: StatsD
+
+    constructor(db: DB, statsd?: StatsD) {
+        this.db = db
+        this.statsd = statsd
+        this.eventPropertiesBuffer = { totalVolume: 0, uniqueVolume: 0, buffer: new Map() }
+        this.lastFlushAt = DateTime.now()
+    }
+
+    public async updateEventPropertyCounter(teamId: number, event: string, properties: Properties): Promise<void> {
+        this.updateEventPropertiesBuffer(teamId, event, properties)
+        // Flush every 2 minutes or 50k unique properties, whichever comes first.
+        // Additionally, a flush is broadcast every 1 minute from pluginServer.
+        if (
+            this.eventPropertiesBuffer.uniqueVolume > 50000 ||
+            DateTime.now().diff(this.lastFlushAt).as('seconds') > 120
+        ) {
+            await this.flush()
+        }
+    }
+
+    /** Save information about event properties into a custom buffer */
+    public updateEventPropertiesBuffer(teamId: number, event: string, properties: Record<string, any>): void {
+        const timestamp = DateTime.now().toSeconds()
+        let bufferForTeam = this.eventPropertiesBuffer.buffer.get(teamId)
+        if (!bufferForTeam) {
+            bufferForTeam = new Map()
+            this.eventPropertiesBuffer.buffer.set(teamId, bufferForTeam)
+        }
+        for (const [property, value] of Object.entries(properties)) {
+            const key = JSON.stringify([event, property])
+            let propertyBuffer = bufferForTeam.get(key)
+            if (!propertyBuffer) {
+                propertyBuffer = {
+                    createdAt: timestamp,
+                    lastSeenAt: timestamp,
+                    totalVolume: 1,
+                    propertyType: null,
+                    propertyTypeFormat: null,
+                }
+                bufferForTeam.set(key, propertyBuffer)
+                this.eventPropertiesBuffer.uniqueVolume += 1
+            } else {
+                propertyBuffer.createdAt = Math.min(timestamp, propertyBuffer.createdAt)
+                propertyBuffer.lastSeenAt = Math.max(timestamp, propertyBuffer.lastSeenAt)
+                propertyBuffer.totalVolume += 1
+            }
+            this.eventPropertiesBuffer.totalVolume += 1
+
+            propertyBuffer.propertyType =
```

Comment:
What happens when we have outlier property values? e.g. I have a bug that every once in a while I send an invalid value (e.g. for a timestamp)? Can we base this based on the format of the majority instead of the latest?